flink-runtime/src/main/java/org/apache/flink/runtime/state/TaskExecutorStateChangelogStoragesManager.java

```diff
@@ -0,0 +1,157 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.state;
+
+import org.apache.flink.api.common.JobID;
+import org.apache.flink.configuration.Configuration;
+import org.apache.flink.runtime.state.changelog.StateChangelogStorage;
+import org.apache.flink.runtime.state.changelog.StateChangelogStorageLoader;
+import org.apache.flink.util.ShutdownHookUtil;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nonnull;
+import javax.annotation.concurrent.GuardedBy;
+
+import java.util.HashMap;
+import java.util.Map;
+
+/** This class holds the all {@link StateChangelogStorage} objects for a task executor (manager). */
+public class TaskExecutorStateChangelogStoragesManager {
+
+    /** Logger for this class. */
+    private static final Logger LOG =
+            LoggerFactory.getLogger(TaskExecutorStateChangelogStoragesManager.class);
+
+    /**
+     * This map holds all state changelog storages for tasks running on the task manager / executor
+     * that own the instance of this. Maps from job id to all the subtask's state changelog
+     * storages.
+     */
+    @GuardedBy("lock")
+    private final Map<JobID, StateChangelogStorage<?>> changelogStoragesByJobId;
+
+    /** Guarding lock for changelogStoragesByJobId and closed-flag. */
+    private final Object lock;
+
+    @GuardedBy("lock")
+    private boolean closed;
+
+    /** shutdown hook for this manager. */
+    private final Thread shutdownHook;
+
+    public TaskExecutorStateChangelogStoragesManager() {
+        this.changelogStoragesByJobId = new HashMap<>();
+        this.lock = new Object();
+        this.closed = false;
+
+        // register a shutdown hook
+        this.shutdownHook =
+                ShutdownHookUtil.addShutdownHook(this::shutdown, getClass().getSimpleName(), LOG);
+    }
+
+    public StateChangelogStorage<?> stateChangelogStorageForJob(
+            @Nonnull JobID jobId, Configuration configuration) {
+        synchronized (lock) {
+            if (closed) {
```

Comment:
Why do we need to synhronize here? Is it because of `closed` flag?
I'm asking because the task thread isn't yet created when this method is called;
In `TaskExecutor.submitTask` (which calls this method), job and taskSlot tables are accessed without any synchronization.

Reply:
Good point.

This is mainly about the ```closed``` and ```changelogStoragesByJobId``` synchronizing among ```stateChangelogStorageForJob```, ```releaseStateChangelogStorageForJob``` and ```shutdown```. I'm not very familiar with the thread model here, but it seems this manager will be accessed same time with the job table. So I think the lock can be removed.

Reply:
I have removed this lock.